### PR TITLE
CMake: add HIP support

### PR DESCRIPTION
**Pull request by mic84**
## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

### CODE REVIEW

File: Tools/CMake/AMReXParallelBackends.cmake [hunk 1]
```diff
@@ -64,3 +64,34 @@ if (ENABLE_DPCPP)
...
+      if(NOT DEFINED ENV{HIP_PATH})
+         set(HIP_PATH "/opt/rocm/hip" CACHE PATH "Path to which HIP has been installed")
+      else()
+         set(HIP_PATH $ENV{HIP_PATH} CACHE PATH "Path to which HIP has been installed")
+      endif()
+   endif()
+
+   set(CMAKE_MODULE_PATH "${HIP_PATH}/cmake" ${CMAKE_MODULE_PATH})
+
+   find_package(HIP REQUIRED)
```
Comment (ax3l) [inline Tools/CMake/AMReXParallelBackends.cmake:97]: We only target the `hip-clang` compiler, which uses the same [compilation model](https://llvm.org/docs/CompileCudaWithLLVM.html#compilation-models) as `clang -x cuda`:
```suggestion
   find_package(HIP REQUIRED)
   if("${HIP_COMPILER}" STREQUAL "hcc")
      message(FATAL_ERROR "Using (deprecated) HCC compiler: please update ROCm")
   endif()
```

[Environment hints](https://rocmdocs.amd.com/en/latest/Programming_Guides/HIP-FAQ.html?highlight=HIP_COMPILER#hip-detected-my-platform-hip-clang-vs-nvcc-incorrectly-what-should-i-do) we might have to document:
```
export HIP_COMPILER=clang
export HIP_PLATFORM=hcc  # or: nvcc
```

File: Tools/CMake/AMReXParallelBackends.cmake [hunk 2]
```diff
@@ -64,3 +64,34 @@ if (ENABLE_DPCPP)
...
+
+   if(HIP_FOUND)
+      message(STATUS "Found HIP: " ${HIP_VERSION})
```
Comment (ax3l) [inline Tools/CMake/AMReXParallelBackends.cmake]:
```suggestion
      message(STATUS "Found HIP: ${HIP_VERSION}")
      message(STATUS "HIP: Platform=${HIP_PLATFORM} Compiler=${HIP_COMPILER}")
```

File: .github/workflows/dependencies/dependencies_hip.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#
+# @result 0 if no files are found, else 1
+#
+
+set -eu -o pipefail
+
+# Ref.: https://rocmdocs.amd.com/en/latest/Installation_Guide/Installation-Guide.html#ubuntu
+wget -q -O - http://repo.radeon.com/rocm/rocm.gpg.key \
+  | sudo apt-key add -
+echo 'deb [arch=amd64] http://repo.radeon.com/rocm/apt/debian/ xenial main' \
```
Comment (ax3l) [inline .github/workflows/dependencies/dependencies_hip.sh:18]: "xenial" is confusingly named: https://github.com/RadeonOpenCompute/ROCm/issues/1207
The "xenial" packages in ROCm 3.7 work only with bionic and focal. There are no aliases yet for bionic and focal.

File: .github/workflows/linux.yml [hunk 4]
```diff
@@ -179,6 +179,29 @@ jobs:
             -DCMAKE_Fortran_COMPILER=$(which gfortran)
         make -j 2 tutorials
 
+  tutorials-hip:
+    name: HIP ROCm@3.7 GFortran@7.5 C++17 [tutorials]
```
Comment (ax3l) [inline .github/workflows/linux.yml]:
```suggestion
    name: HIP ROCm@3.7 GFortran@9.3 C++17 [tutorials]
```

File: Tools/CMake/AMReXTargetHelpers.cmake [hunk 5]
```diff
@@ -122,3 +122,40 @@ function (setup_target_for_cuda_compilation _target)
...
+      HIPCC_OPTIONS ${_hipcc_options}
+      HCC_OPTIONS ${_hcc_options}
+      NVCC_OPTIONS ${_nvcc_options})
+
+   if (_source_files)
+      list(REMOVE_ITEM _sources ${_source_files})
+   endif ()
+
+   # overwrite sources of _target with "new" sources
+   set_target_properties(${_target} PROPERTIES SOURCES ${_generated_files} ${_sources})
```
Comment (ax3l) [inline Tools/CMake/AMReXTargetHelpers.cmake]:
```
  set_target_properties called with incorrect number of arguments.
```
```suggestion
   set_target_properties(${_target} PROPERTIES SOURCES "${_generated_files};${_sources}")
```

Refs.:
- https://stackoverflow.com/a/32180996/2719194

File: Tools/CMake/AMReXTargetHelpers.cmake [hunk 6]
```diff
@@ -122,3 +122,40 @@ function (setup_target_for_cuda_compilation _target)
...
+      HIPCC_OPTIONS ${_hipcc_options}
+      HCC_OPTIONS ${_hcc_options}
+      NVCC_OPTIONS ${_nvcc_options})
+
+   if (_source_files)
+      list(REMOVE_ITEM _sources ${_source_files})
+   endif ()
+
+   # overwrite sources of _target with "new" sources
+   set_target_properties(${_target} PROPERTIES SOURCES "${_generated_files};${_sources}")
```
Comment (ax3l) [inline Tools/CMake/AMReXTargetHelpers.cmake]: Should those operate on `amrex-hip` now?
```suggestion
   set_target_properties(${_target}-hip PROPERTIES SOURCES "${_generated_files};${_sources}")
```

File: Tools/CMake/AMReXTargetHelpers.cmake [hunk 7]
```diff
@@ -122,3 +122,40 @@ function (setup_target_for_cuda_compilation _target)
...
+
+   # set linker language
+   set_target_properties(${_target} PROPERTIES LINKER_LANGUAGE ${HIP_C_OR_CXX})
```
Comment (ax3l) [inline Tools/CMake/AMReXTargetHelpers.cmake]: Maybe:
```suggestion
   set_target_properties(${_target}-hip PROPERTIES LINKER_LANGUAGE ${HIP_C_OR_CXX})
```

File: Tools/CMake/AMReXTargetHelpers.cmake [hunk 8]
```diff
@@ -122,3 +122,40 @@ function (setup_target_for_cuda_compilation _target)
...
+   set(_hcc_options)
+   set(_nvcc_options)
+   hip_prepare_target_commands(${_target}
+      OBJ _generated_files _source_files ${_sources} ${_cmake_options}
+      HIPCC_OPTIONS ${_hipcc_options}
+      HCC_OPTIONS ${_hcc_options}
+      NVCC_OPTIONS ${_nvcc_options})
+
+   if (_source_files)
+      list(REMOVE_ITEM _sources ${_source_files})
```
Comment (ax3l) [inline Tools/CMake/AMReXTargetHelpers.cmake]: Not 100% sure if `list(REMOVE_ITEM` supports `<list> <list>`. The docs say `<list> <value> [<value> ...]` as arguments. Also relative/absolute paths could be good to double-check here.

File: Tools/CMake/AMReX_Defines.cmake [hunk 9]
```diff
@@ -67,6 +67,7 @@ function ( set_amrex_defines )
 
    # HIP
    add_amrex_define( AMREX_USE_HIP NO_LEGACY IF ENABLE_HIP )
+   add_amrex_define( NDEBUG IF ENABLE_HIP)  # This address a bug that causes slow build times
```
Comment (ax3l) [inline Tools/CMake/AMReX_Defines.cmake:70]:
```suggestion
   add_amrex_define( NDEBUG IF ENABLE_HIP)  # This address a bug that causes slow build times (seen first with ROCm 3.7)
```

File: .github/workflows/linux.yml [hunk 10]
```diff
@@ -3,234 +3,260 @@ name: linux
...
-      run: |
-        mkdir build
-        cd build
-        cmake ..                                  \
-            -DCMAKE_VERBOSE_MAKEFILE=ON           \
-            -DCMAKE_INSTALL_PREFIX=/tmp/my-amrex  \
-            -DCMAKE_CXX_STANDARD=17
-        make -j 2
-        make install
+  # # Build and install libamrex as AMReX CMake project
```
Comment (ax3l) [inline .github/workflows/linux.yml]: memo to ourselves: don't forget to un-comment again

File: Tutorials/CMakeLists.txt [hunk 11]
```diff
@@ -94,12 +94,23 @@ function (setup_tutorial _srcs  _inputs)
...
    if (${_inputs})
       file( COPY ${${_inputs}} DESTINATION ${_exe_dir} )
    endif ()
 
 endfunction ()
 
+#
+# Find HIP
+# HIP stuff are not transitive, i.e. will not be inherited from amrex.
+# Also, HIP is not yet a firt-class language in CMake, like CUDA.
```
Comment (ax3l) [inline Tutorials/CMakeLists.txt]:
```suggestion
# Also, HIP is not yet a first-class language in CMake, like CUDA.
```

File: .github/workflows/linux.yml [hunk 12]
```diff
@@ -3,223 +3,249 @@ name: linux
...
+  #           -DENABLE_TUTORIALS=ON                          \
+  #           -DENABLE_PARTICLES=ON                          \
+  #           -DENABLE_DPCPP=ON                              \
+  #           -DCMAKE_C_COMPILER=$(which clang)              \
+  #           -DCMAKE_CXX_COMPILER=$(which dpcpp)            \
+  #           -DCMAKE_Fortran_COMPILER=$(which gfortran)
+  #       make -j 2 tutorials
+
+  tutorials-hip:
+    name: HIP ROCm@3.7 GFortran@9.3 C++17 [tutorials]
```
Comment (WeiqunZhang) [inline .github/workflows/linux.yml]: I think we should remove `@3.7` from the name.  It's already outdated.  The CI now runs with ROCm 3.8.

Comment (mic84) [reply .github/workflows/linux.yml]: Done

File: Tools/CMake/AMReXParallelBackends.cmake [hunk 13]
```diff
@@ -64,3 +64,60 @@ if (ENABLE_DPCPP)
...
+   # Link to hiprand -- must include rocrand too
+   find_package(rocrand REQUIRED CONFIG)
+   find_package(hiprand REQUIRED CONFIG)
+   target_link_libraries(amrex PUBLIC hip::hiprand roc::rocrand)
+
+   # ARCH flags -- these must be PUBLIC for all downstream targets to use,
+   # else there will be a runtime issue (cannot find
+   # missing gpu devices)
+   target_compile_options(amrex
+      PRIVATE
```
Comment (sayerhs) [inline Tools/CMake/AMReXParallelBackends.cmake]: @mic84 Should this be PUBLIC here, as the comment above says? In my test builds of AMR-Wind on Tulip, I am finding that `--amdgpu-target` is not being set for application targets. I had to manually add them to the compilation flags.

Comment (mic84) [reply Tools/CMake/AMReXParallelBackends.cmake]: Yes, my bad. Fixing now

File: Docs/sphinx_documentation/source/GPU.rst [hunk 14]
```diff
@@ -302,6 +302,55 @@ we provide the helper function ``setup_target_for_cuda_compilation()``:
    # Link against amrex
    target_link_libraries(my_target AMReX::amrex)
 
+
+
+Enabling HIP support (experimental)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To build AMReX with HIP support in CMake, add
+``-DENABLE_CUDA=YES -DAMD_ARCH=<target-arch> -DCMAKE_CXX_COMPILER=<your-hip-compiler>`` to the
```
Comment (sayerhs) [inline Docs/sphinx_documentation/source/GPU.rst]:
```suggestion
``-DENABLE_HIP=YES -DAMD_ARCH=<target-arch> -DCMAKE_CXX_COMPILER=<your-hip-compiler>`` to the
```

**Review by WeiqunZhang — APPROVED** [review]

**Pull request merged**